### PR TITLE
Set groundwork for poc-v5 and fix scoring

### PR DESCRIPTION
**Pull request by vihu**
Add witness quality checks when scoring. Will update this further.

### CODE REVIEW

File: src/transactions/v1/blockchain_txn_poc_receipts_v1.erl [hunk 1]
```diff
@@ -297,7 +299,23 @@ deltas(Txn) ->
                                    Witnesses = blockchain_poc_path_element_v1:witnesses(Element),
                                    NextElements = lists:sublist(Path, N+1, Length),
                                    HasContinued = check_path_continuation(NextElements),
-                                   {Val, Continue} = assign_alpha_beta(HasContinued, Receipt, Witnesses),
+                                   {Val, Continue} = assign_alpha_beta(HasContinued, Challengee, Receipt, Witnesses, undefined, undefined),
```
Comment (Vagabond) [inline src/transactions/v1/blockchain_txn_poc_receipts_v1.erl]: Why is there another function clause if you just pass undefined, undefined to the same function?

Comment (vihu) [reply src/transactions/v1/blockchain_txn_poc_receipts_v1.erl]: Right, don't know what I was thinking. Check: edadab1

File: src/transactions/v1/blockchain_txn_poc_receipts_v1.erl [hunk 2]
```diff
@@ -341,19 +362,116 @@ assign_alpha_beta(HasContinued, Receipt, Witnesses) ->
...
+                                             Challengee :: libp2p_crypto:pubkey_bin(),
+                                             Witnesses :: [blockchain_poc_witness_v1:poc_witness()],
+                                             Ledger :: blockchain_ledger_v1:ledger()) -> {float(), 0}.
+update_alpha_beta_with_witness_quality(undefined, _POCVersion, Challengee, Witnesses, Ledger) ->
+    %% no poc receipt
+    WitnessQualityChecks = witness_quality_checks(Challengee, Witnesses, Ledger),
+    case lists:all(fun(C) -> C == true end, WitnessQualityChecks) of
+        false ->
+            %% Either the witnesses are too close
+            %% or the RSSIs are too low
```
Comment (Vagabond) [inline src/transactions/v1/blockchain_txn_poc_receipts_v1.erl]:
```suggestion
            %% or the RSSIs are too high
```

File: src/transactions/v1/blockchain_txn_poc_receipts_v1.erl [hunk 3]
```diff
@@ -341,19 +362,116 @@ assign_alpha_beta(HasContinued, Receipt, Witnesses) ->
...
+            %% high alpha bump, but not as high as when there is a receipt
+            {0.7, 0}
+    end;
+update_alpha_beta_with_witness_quality(_Receipt, _POCVersion, Challengee, Witnesses, Ledger) ->
+    %% poc receipt
+    WitnessQualityChecks = witness_quality_checks(Challengee, Witnesses, Ledger),
+    case lists:all(fun(C) -> C == true end, WitnessQualityChecks) of
+        false ->
+            %% Either the witnesses are too close
+            %% or the RSSIs are too low
```
Comment (Vagabond) [inline src/transactions/v1/blockchain_txn_poc_receipts_v1.erl]:
```suggestion
            %% or the RSSIs are too high
```

File: src/blockchain_poc_path_v2.erl [hunk 4]
```diff
@@ -318,42 +304,85 @@ check_witness_distance(WitnessParent, ParentIndices, ExclusionCells) ->
...
+                Hist ->
+                    case maps:get(28, Hist, 0) of
+                        0 ->
+                            %% No bad RSSIs found, include
+                            true;
+                        BadCount ->
+                            %% If the bad RSSI count does not dominate
+                            %% the overall RSSIs this witness has, include,
+                            %% otherwise exclude
+                            BadCount < lists:sum(maps:values(Hist))
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/blockchain_poc_path_v2.erl]: A re-assert would clear them

[the remark this answers is not in the document]
Comment (vihu) [reply src/blockchain_poc_path_v2.erl]: I also think that eventually the sum of all rssi values would overcome the supposedly impossible RSSIs, given that the hotspot either reasserts or is moved or just starts getting viable neighbors?

File: src/blockchain_poc_path_v2.erl [hunk 5]
```diff
@@ -318,42 +304,85 @@ check_witness_distance(WitnessParent, ParentIndices, ExclusionCells) ->
...
+        V when is_integer(V), V > 4 ->
+            maps:is_key(WitnessPubkeyBin, ActiveGateways);
+        _ ->
+            true
+    end.
+
+-spec rssi_weight(Vars :: map()) -> undefined | float().
 rssi_weight(Vars) ->
-    maps:get(poc_v4_prob_rssi_wt, Vars, ?POC_V4_PROB_RSSI_WT).
+    maps:get(poc_v4_prob_rssi_wt, Vars, undefined).
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/blockchain_poc_path_v2.erl]: Yes, just use maps:get/2 which blows up if unset

[the remark this answers is not in the document]
Comment (vihu) [reply src/blockchain_poc_path_v2.erl]: Ok, testing a resync locally. Will update.

**Pull request merged**